packages/core/src/utils/indyIdentifiers.ts

```diff
@@ -0,0 +1,91 @@
+import type { CredDef, Schema } from 'indy-sdk'
+
+import { AriesFrameworkError } from '../error/AriesFrameworkError'
+import { generateCredentialDefinitionId, generateSchemaId } from '../modules/ledger/ledgerUtil'
+
+import { didFromCredentialDefinitionId, didFromSchemaId } from './did'
+
+/**
+ *
+ * @see For the definitions below see also: https://hyperledger.github.io/indy-did-method/#indy-did-method-identifiers
+ *
+ */
+export type Did = 'did'
+export type DidIndyMethod = 'indy'
+// Maybe this can be typed more strictly than string. Choosing string for now as this can be eg just `sovrin` or eg `sovrin:staging`
+export type DidIndyNamespace = string
+//  NOTE: because of the ambiguous nature - whether there is a colon or not within DidIndyNamespace this is the substring after the ***last*** colon
+export type NamespaceIdentifier = string
+
+// TODO: This template literal type can possibly be improved. This version leaves the substrings as potentially undefined
+export type IndyNamespace = `${Did}:${DidIndyMethod}:${DidIndyNamespace}:${NamespaceIdentifier}`
+
+export function isQualifiedIdentifier(identifier: string | undefined): boolean {
+  if (!identifier || identifier === '') return false
+  return identifier.startsWith('did:indy:')
+}
+
+export function unqualifyIndyDid(qualifiedIdentifier: string): string {
+  if (!isQualifiedIdentifier(qualifiedIdentifier)) return qualifiedIdentifier
+
+  const lastColonIndex = qualifiedIdentifier.lastIndexOf(':')
+  const identifierTrunk = qualifiedIdentifier.substring(lastColonIndex + 1)
+  const txType = identifierTrunk.split('/')[3]
+
+  if (txType === 'SCHEMA') {
+    // did:indy:sovrin:F72i3Y3Q4i466efjYJYCHM/anoncreds/v0/SCHEMA/npdb/4.3.4 -> F72i3Y3Q4i466efjYJYCHM:2:npdb:4.3.4
+    const [id, , , , name, version] = identifierTrunk.split('/')
+    return generateSchemaId(id, name, version)
+  } else if (txType === 'CLAIM_DEF') {
+    // did:indy:sovrin:5nDyJVP1NrcPAttP3xwMB9/anoncreds/v0/CLAIM_DEF/56495/npdb -> 5nDyJVP1NrcPAttP3xwMB9:3:CL:56495:npbd
+    const [id, , , , seqNo, name] = identifierTrunk.split('/')
+    return generateCredentialDefinitionId(id, +seqNo, name)
+  } else {
+    // indy Union Indy Ledger w/o url syntax
+    // did:indy:idunion:test:2MZYuPv2Km7Q1eD4GCsSb6 -> 2MZYuPv2Km7Q1eD4GCsSb6
+    return identifierTrunk
+  }
+}
+
+/**
+ *
+ * @see https://hyperledger.github.io/indy-did-method/#schema
+ *
+ */
+export function getDidUrlTrunkFromSchema(schema: Schema): string {
+  const namespaceIdentifier = didFromSchemaId(schema.id)
+  const didUrl = `${namespaceIdentifier}/anoncreds/v0/SCHEMA/${schema.name}/${schema.version}`
```

Comment:
The schema id should already contain this part or it should be a legacy identifier, in which case it won't include the indy namspace identifier. I think this state we're assuming here is invalid

Reply:
hmmm ok. Let me have a look at this. Personally unsure right now but you're probably right

Reply:
Not sure I fully get your remark here. As per https://hyperledger.github.io/indy-did-method/#schema this gets `F72i3Y3Q4i466efjYJYCHM` from `F72i3Y3Q4i466efjYJYCHM:2:npdb:4.3.4` and creates the 'trunk' (`F72i3Y3Q4i466efjYJYCHM/anoncreds/v0/SCHEMA/npdb/4.3.4`) to be added to `did:indy:sovrin` creating the entire `did:indy:sovrin:F72i3Y3Q4i466efjYJYCHM/anoncreds/v0/SCHEMA/npdb/4.3.4`

What am I missing/misunderstanding here?

Reply:
What @TimoGlastra means, I believer, is that the `schema.id` should already have this `didUrl` part as it is the schema id.

A legacy schema id does have the `namespace` in the `schema.id` so we can never extract this as is done in line 56.

Also, `didFromSchemaId(schema.id)`, for legacy and qualified, will never return a namespace idenfitier as it just does `"5nDyJVP1NrcPAttP3xwMB9:3:CL:56495:npdb".split(":")[0]` which returns the DID ONLY for unqualified identifiers.

Reply:
Ah I think I misunderstood the terms here. namespaceIdentifier IS the did, correct?

Reply:
yes in this case! Maybe i should rename it did. Just followed the naming from did:indy:<namespace>:<namespace identifier> (also in https://hyperledger.github.io/indy-did-method/#indy-did-method-identifiers, bottom of the section)